### PR TITLE
DS-199. Authentication and Authorization in Generic Geosearch

### PR DESCRIPTION
**Pull request by ngaranko**
Adding Authentication and Authorization to Generic Geosearch endpoint.
Additional endpoint `/catalogus/` added, listing `datasets` available for searching to given user.

### CODE REVIEW

File: web/geosearch/tests/test_catalogus_endpoint.py [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+import flask
+import pytest
+
+from datapunt_geosearch import config, create_app
+from datapunt_geosearch.registry import registry
+
+
+@pytest.mark.usefixtures("dataservices_db")
+@pytest.mark.usefixtures("dataservices_fake_data")
+@pytest.mark.usefixtures("create_authz_token")
```
Comment (vdboor) [inline web/geosearch/tests/test_catalogus_endpoint.py]: mag ook als 1 regel

Comment (vdboor) [inline web/geosearch/tests/test_catalogus_endpoint.py]:
```suggestion
@pytest.mark.usefixtures("dataservices_db", "dataservices_fake_data", "create_authz_token")
```

File: web/geosearch/datapunt_geosearch/datasource.py [hunk 2]
```diff
@@ -50,6 +50,12 @@ def __init__(self, dsn=None, connection=None):
 
         self.meta = self.metadata.copy()
 
+    @classmethod
+    def check_scopes(cls, scopes=None):
```
Comment (vdboor) [inline web/geosearch/datapunt_geosearch/datasource.py]:
```suggestion
    def check_scopes(cls, scopes=None):
        """Check who may access this datasource. When no scopes are given,
        access is allowed when there are no restrictions to the data.
        """
```

Also, perhaps rename `check_scopes()` into something like `is_authorized()`?

File: .jenkins/runtests.sh [hunk 3]
```diff
@@ -29,7 +29,7 @@ echo "let's see if databases are working"
 # load latest data into databases
 # TODO only tables!
 dc exec -T nap_db /bin/update-db.sh nap
-dc exec -T bag_v11_db /bin/update-db.sh bag_v11
+dc exec -T bag_v11_db /bin/update-db.sh bag_v11 || true
```
Comment (vdboor) [inline .jenkins/runtests.sh]: Perhaps add a comment why this is done?

File: web/geosearch/datapunt_geosearch/config.py [hunk 4]
```diff
@@ -131,3 +134,29 @@ def get_db_settings(db: str, docker_host: str, localport: str) -> Dict[str, str]
...
+                filename,
+                repr(error_details)
+            )
+            return None, None, None
+
+    return config.get('jwks'), config.get('jwks_url'), config.get('jwks_allowed_signing_algorithms')
+
+
+default_config_file = os.path.join(os.path.dirname(__file__), 'config.yaml')
+JWKS, JWKS_URL, JWKS_SIGNING_ALGORITHMS = load_jwks_config(default_config_file)
```
Comment (vdboor) [inline web/geosearch/datapunt_geosearch/config.py]: This is actually an if-else: if `/etc/geosearch.yaml` is found, this code is completely ignored.

File: web/geosearch/datapunt_geosearch/authz.py [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+    """
+    Optionally check authentication via on request.
+
+    Returns None if request contains no Authorization header
+     or header is correct.
+    In case Authorization header is correct - updates `g.authz_scopes` with
+     scopes from claims defined in JWT token.
+    Aborts with 401 in case Authorization header contains incorrect token.
+    """
+    g.authz_scopes = None
```
Comment (vdboor) [inline web/geosearch/datapunt_geosearch/authz.py]: Questions: using `request.authz_scopes` is not a proper Flask practice?

File: web/geosearch/datapunt_geosearch/blueprints/search.py [hunk 6]
```diff
@@ -76,14 +80,26 @@ def search_everywhere():
...
         y=y,
         rd=rd,
-        limit=limit
+        limit=limit,
     ))
 
     return Response(generate_async(
-        request_args=request_args
+        request_args=request_args,
+        authz_scopes=getattr(g, "authz_scopes", None)
```
Comment (vdboor) [inline web/geosearch/datapunt_geosearch/blueprints/search.py]: For something as important as authorization, perhaps have a global function instead? e.g. `get_current_authz_scopes()` ?

File: web/geosearch/datapunt_geosearch/config.py (hunk 4 above)
Comment (ngaranko) [reply web/geosearch/datapunt_geosearch/config.py]: This is absolutely correct, as `/etc/geosearch.yaml` is environment specific config and should overwrite default, if exists.

File: web/geosearch/datapunt_geosearch/blueprints/search.py (hunk 6 above)
Comment (ngaranko) [reply web/geosearch/datapunt_geosearch/blueprints/search.py]: Yeah, good idea.

File: web/geosearch/datapunt_geosearch/authz.py (hunk 5 above)
Comment (ngaranko) [reply web/geosearch/datapunt_geosearch/authz.py]: In current use case there's no difference, however if we intend to use this code outside of request - using `g` will be much better approach. Therefore use of `g` is future proof solution, and personal preference :)

**Pull request closed without merging**